app/models/foreign_input_set.rb

```diff
@@ -0,0 +1,45 @@
+class ForeignInputSet < ActiveRecord::Base
+  class CircularDependencyError < Foreman::Exception
+  end
+
+  TYPES = { :user => N_('User input'), :fact => N_('Fact value'), :variable => N_('Variable'),
+            :puppet_parameter => N_('Puppet parameter') }.with_indifferent_access
+
+  attr_accessible :name, :template_id, :target_template_id, :include_all, :include, :exclude
+
+  belongs_to :template
+  belongs_to :target_template, :class_name => 'Template'
+
+  validates :target_template, :presence => true
+  validate :check_circular_dependency
+
+  def inputs(templates_stack = [])
+    if templates_stack.include?(target_template)
+      raise CircularDependencyError.new(N_("Circular dependency detected in foreign input set '%{template}' -> '%{target_template}'. Templates stack: %{templates_stack}"),
+                                        :template => template.name, :target_template => target_template.name, :templates_stack => templates_stack.map(&:name).inspect)
+    end
+    inputs = target_template.template_inputs_with_foreign(templates_stack + [target_template])
+    unless self.include_all?
+      inputs = inputs.select { |input| included_names.include?(input.name) }
+    end
+    inputs = inputs.reject { |input| excluded_names.include?(input.name) }
+    return inputs
+  end
+
+  def included_names
+    self.include.to_s.split(',').map(&:strip)
+  end
+
+  def excluded_names
+    self.exclude.to_s.split(',').map(&:strip)
```

Comment:
could you please dry this up with included_names? some private method would do...